### PR TITLE
On demand maps

### PR DESCRIPTION
**Pull request by sei-byoung**
Implement the ability to subscribe to events in Player. Apps can supply the URL to a callback, which will be called when the relevant event occurs.

### CODE REVIEW

**Review by sei-ebram — APPROVED** [review]

File: Player.Api.Data/Data/Models/View.cs [hunk 1]
```diff
@@ -28,11 +27,13 @@ public class ViewEntity
         public ViewEntity Clone()
         {
             var entity = this.MemberwiseClone() as ViewEntity;
-            entity.Id = Guid.Empty;
+            entity.Id = Guid.NewGuid();
```
Comment (sei-aschlackman) [inline Player.Api.Data/Data/Models/View.cs:30]: Does it work without this after the transaction interceptor change?

File: Player.Api/Controllers/WebhookController.cs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        [HttpDelete("webhooks/{id}")]
+        [ProducesResponseType((int) HttpStatusCode.NoContent)]
+        [SwaggerOperation(OperationId = "delete")]
+        public async Task<IActionResult> Delete([FromRoute] Guid id, CancellationToken ct)
+        {
+            await _webhookService.DeleteAsync(id, ct);
+            return NoContent();
+        }
+    }
+}
```
Comment (sei-aschlackman) [inline Player.Api/Controllers/WebhookController.cs:66]: Is there no way to update an existing subscription? For example, to add or remove subscribed event types?

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+                    AddEvent(t);
+                }
+            }
+        }
+
+        public void AddEvent(Task t)
+        {
+            // Should this use SendAsync? 
+            var success = _eventQueue.Post(t);
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:70]: I think this should use SendAsync, to answer the comment.

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+        private readonly IHttpClientFactory _clientFactory; 
+        private ActionBlock<Task> _eventQueue;
+        
+
+        public BackgroundWebhookService(ILogger<BackgroundWebhookService> logger, IServiceScopeFactory scopeFactory, IHttpClientFactory clientFactory)
+        {
+            _logger = logger;
+            _scopeFactory = scopeFactory;
+            _clientFactory = clientFactory;
+            _eventQueue = new ActionBlock<Task>(t => t.Start());
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:45]: I think it makes more sense for this to accept an event object or id and have the ActionBlock call ProcessEvent. Passing in an arbitrary Task is a little too generic and puts a lot on the caller of the method.

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                        foreach (var sub in subs)
+                        {
+                            HttpResponseMessage resp = null;
+                            var payload = new ViewModels.Webhooks.ViewCreated();
+                            payload.ViewId = createdView.Id;
+                            payload.ParentId = createdView.ViewId != null ? (Guid) createdView.ViewId : Guid.Empty;
+                            
+                            var webhookEvent = new ViewModels.Webhooks.WebhookEvent();
+                            webhookEvent.Name = "View Created";
+                            webhookEvent.Timestamp = DateTime.Now;
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:108]: This should be DateTime.UtcNow

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+                            .SingleOrDefaultAsync();
+
+                        foreach (var sub in subs)
+                        {
+                            HttpResponseMessage resp = null;
+                            var payload = new ViewModels.Webhooks.ViewCreated();
+                            payload.ViewId = createdView.Id;
+                            payload.ParentId = createdView.ViewId != null ? (Guid) createdView.ViewId : Guid.Empty;
+                            
+                            var webhookEvent = new ViewModels.Webhooks.WebhookEvent();
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:106]: Would it be possible for the WebhookEvent to be created by the event handler and added as a serialized object to the PendingEvent in the database, so the background worker can just pull it out and send it and not have to be updated for every new event type?

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+        private async Task<HttpResponseMessage> SendJsonPost(string json, string url, string auth)
+        {
+            using (var client = _clientFactory.CreateClient())
+            {
+                client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", auth);
+                return await client.PostAsync(url, new StringContent(json, Encoding.UTF8, "application/json"));
+            }
+        }
+
+        private async Task<string> getAuthToken(string clientId, string clientSecret, string tokenUrl)
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:209]: This method could maybe be simplified a bit by using the [IdentityModel](https://www.nuget.org/packages/IdentityModel/) package similar to [in Caster here](https://github.com/cmu-sei/Caster.Api/blob/development/src/Caster.Api/Domain/Services/AuthenticationService.cs#L79)

**Review by sei-tspencer — APPROVED** [review]

File: Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+        {
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/BackgroundServices/BackgroundWebhookService.cs:210]: We should also consider caching the token per client, but that can be a separate PR

File: Player.Api/Infrastructure/DbInterceptors/ViewHandler.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            PlayerContext context, 
+            ILogger<ViewHandlerBase> logger,
+            IBackgroundWebhookService backgroundService) : base(context, logger, backgroundService) {}
+
+        public async Task Handle(EntityCreated<ViewEntity> notification, CancellationToken ct)
+        {
+            // Add pending event to db
+            var eventEntity = new PendingEventEntity();
+            eventEntity.EventType = EventType.ViewCreated;
+            eventEntity.Timestamp = DateTime.Now;
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/DbInterceptors/ViewHandler.cs:42]: This should be DateTime.UtcNow

File: Player.Api/Infrastructure/DbInterceptors/ViewHandler.cs [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            eventEntity.EffectedEntityId = notification.Entity.Id;
+
+            _context.Add(eventEntity);
+            await _context.SaveChangesAsync(ct);
+
+            // Add event to event queue
+            Task t = new Task(async eventId => await _backgroundService.ProcessEvent((Guid) eventId), eventEntity.Id, new CancellationToken());
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/DbInterceptors/ViewHandler.cs:49]: Like I said above, this is kind of clunky. Would be nice to just do something like _backgroundService.AddEvent(event) or AddEvent(eventId).

File: Player.Api/Infrastructure/Extensions/DatabaseExtensions.cs [hunk 11]
```diff
@@ -30,6 +33,7 @@ public static IWebHost InitializeDatabase(this IWebHost webHost)
                     var databaseOptions = services.GetService<DatabaseOptions>();
                     var ctx = services.GetRequiredService<PlayerContext>();
                     var seedDataOptions = services.GetService<SeedDataOptions>();
+                    var backgroundService = services.GetService<IBackgroundWebhookService>();
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/Extensions/DatabaseExtensions.cs:36]: Why is this here?

File: Player.Api/Infrastructure/Mappings/ViewProfile.cs [hunk 12]
```diff
@@ -16,7 +16,8 @@ public class ViewProfile : AutoMapper.Profile
         public ViewProfile()
         {
             CreateMap<ViewEntity, View>()
-                .ForMember(dest => dest.CanManage, opt => opt.MapFrom<ManageViewResolver>());
+                .ForMember(dest => dest.CanManage, opt => opt.MapFrom<ManageViewResolver>())
+                .ForMember(dest => dest.ParentViewId, opt => opt.MapFrom(src => src.ViewId));
```
Comment (sei-aschlackman) [inline Player.Api/Infrastructure/Mappings/ViewProfile.cs:20]: Why not have these names match? ParentViewId on the entity as well probably makes sense?

File: Player.Api/Services/WebhookService.cs [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+namespace Player.Api.Services
+{
+    public interface IWebhookService
+    {
+        Task<WebhookSubscription> Subscribe(WebhookSubscriptionForm form, CancellationToken ct);
+        Task<IEnumerable<WebhookSubscription>> GetAll(CancellationToken ct);
+        Task DeleteAsync(Guid id, CancellationToken ct);
+    }
+
+    public class WebhookService : IWebhookService
```
Comment (sei-aschlackman) [inline Player.Api/Services/WebhookService.cs:28]: Need to add auth checks for these methods?

**Review by sei-aschlackman — CHANGES_REQUESTED** [review]

**Comment by sei-aschlackman** [comment]
Superseeded by https://github.com/cmu-sei/Player.Api/pull/32

**Pull request closed without merging**